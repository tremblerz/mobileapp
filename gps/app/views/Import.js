import React, { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Linking, ScrollView, StyleSheet, View } from 'react-native';

import { Button } from '../components/Button';
import NavigationBarWrapper from '../components/NavigationBarWrapper';
import { Typography } from '../components/Typography';
import colors from '../constants/colors';
import fontFamily from '../constants/fonts';
import { pickFile } from '../helpers/General';
import {
  EmptyFilePathError,
  InvalidFileExtensionError,
  NoRecentLocationsError,
  importTakeoutData,
} from '../helpers/GoogleTakeOutAutoImport';

const makeImportResults = (label = '', error = false) => ({
  error,
  label,
});

const ImportScreen = props => {
  const { t } = useTranslation();
  const {
    navigation: { goBack },
  } = props;
  const [importResults, _setImportResults] = useState(makeImportResults());
  const setImportResults = (...args) =>
    _setImportResults(makeImportResults(...args));
  async function importPickFile() {
    try {
      // reset info message
      setImportResults();

      const filePath = await pickFile();
      if (filePath) {
        await importTakeoutData(filePath);
        setImportResults(t('label.import_success'));
      }
    } catch (err) {
      if (err instanceof NoRecentLocationsError) {
        setImportResults(t('import.google.no_recent_locations'), true);
      } else if (err instanceof InvalidFileExtensionError) {
        setImportResults(t('import.google.invalid_file_format'), true);
      } else if (err instanceof EmptyFilePathError) {
        /**
         * If the imported file is opened from other than Google Drive folder,
         * filepath is returned as null. Leaving a message to ensure import file
         * is located on Google Drive.
         */
        setImportResults(t('import.google.file_open_error'), true);
      } else {
        console.log('[ERROR] Failed to import locations', err);
        setImportResults(t('import.error'), true);
      }
    }
  }

  return (
    <NavigationBarWrapper title={t('import.title')} onBackPress={goBack}>
      <ScrollView style={styles.main}>
        <View style={styles.subHeaderTitle}>
          <Typography style={styles.sectionDescription}>
            {t('import.google.instructions_first')}
          </Typography>
          {/* eslint-disable react/no-unescaped-entities */}
          <Typography style={styles.sectionDescription}>
            {t('import.google.instructions_second')}
          </Typography>
          <Typography style={styles.sectionDescription}>
            {t('import.google.instructions_detailed')}
          </Typography>

          <Button
            small
            label={t('import.google.visit_button_text')}
            testID='google-takeout-link'
            onPress={() =>
              Linking.openURL(
                'https://takeout.google.com/settings/takeout/custom/location_history',
              )
            }
            style={{ marginTop: 24 }}
          />

          <Button
            small
            label={t('import.title')}
            testID='google-takeout-import-btn'
            onPress={importPickFile}
            style={{ marginTop: 24 }}
          />

          {importResults.label ? (
            <Typography
              style={{
                ...styles.importResults,
                ...(importResults?.error ? styles.importResultsError : {}),
              }}>
              {importResults.label}
            </Typography>
          ) : null}
        </View>
      </ScrollView>
    </NavigationBarWrapper>
  );
};

const styles = StyleSheet.create({
  subHeaderTitle: {
    textAlign: 'center',
    fontWeight: 'bold',
    fontSize: 22,
    padding: 5,
    paddingBottom: 20,
  },
  main: {
    flex: 1,
    flexDirection: 'column',
    textAlignVertical: 'top',
    paddingLeft: 20,
    paddingRight: 20,
    width: '100%',
    alignSelf: 'center',
  },
  sectionDescription: {
    fontSize: 16,
    lineHeight: 24,
    marginTop: 12,
    fontFamily: fontFamily.primaryRegular,
  },
  importResults: {
    fontSize: 12,
    lineHeight: 20,
    marginTop: 10,
    textAlign: 'center',
    fontFamily: fontFamily.primaryRegular,
    color: colors.VIOLET_TEXT,
  },
  importResultsError: {
    color: colors.RED_TEXT,
  },
});
export default ImportScreen;
