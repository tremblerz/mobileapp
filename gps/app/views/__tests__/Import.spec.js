import 'react-native';

import {
  NativeTestEvent,
  act,
  fireEvent,
  render,
} from '@testing-library/react-native';
import React from 'react';
import { Linking } from 'react-native';
import renderer from 'react-test-renderer';

import Import from '../Import';

jest.spyOn(console, 'log').mockImplementation(() => {});

const GoogleTakeOutAutoImport = require('../../helpers/GoogleTakeOutAutoImport');
const General = require('../../helpers/General');

const navigationMock = {
  goBack: jest.fn(),
};
let openURLSpy;
let importTakeoutDataSpy;
let pickFileSpy;

describe('Import component', () => {
  beforeAll(() => {
    openURLSpy = jest
      .spyOn(Linking, 'openURL')
      .mockImplementation(() => Promise.resolve());
    importTakeoutDataSpy = jest
      .spyOn(GoogleTakeOutAutoImport, 'importTakeoutData')
      .mockImplementation(() => Promise.resolve([]));
    pickFileSpy = jest
      .spyOn(General, 'pickFile')
      .mockImplementation(() => Promise.resolve('file://tmp/test.zip'));
  });
  afterAll(() => {
    openURLSpy.mockRestore();
    importTakeoutDataSpy.mockRestore();
    pickFileSpy.mockRestore();
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(<Import navigation={navigationMock} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('clicking on Google Takeout button leads to open browser page', () => {
    const { getByTestId } = render(<Import navigation={navigationMock} />);
    fireEvent(getByTestId('google-takeout-link'), new NativeTestEvent('press'));
    expect(openURLSpy).toHaveBeenCalled();
  });

  it('clicking on import Takeout button opens a file picker', async () => {
    const { getByTestId } = render(<Import navigation={navigationMock} />);
    await act(async () => {
      await fireEvent(
        getByTestId('google-takeout-import-btn'),
        new NativeTestEvent('press'),
      );
    });
    expect(pickFileSpy).toHaveBeenCalled();
    expect(importTakeoutDataSpy).toHaveBeenCalled();
  });
});
