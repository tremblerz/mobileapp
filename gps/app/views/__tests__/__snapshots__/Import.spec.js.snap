// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Import component renders correctly 1`] = `
Array [
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "backgroundColor": "#4051DB",
        "flexBasis": 0,
        "flexGrow": 0,
        "flexShrink": 1,
      }
    }
  />,
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "backgroundColor": "#F7F8FF",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#4051DB",
          "borderBottomColor": "#5061E6",
          "borderBottomWidth": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "height": 55,
            "justifyContent": "center",
            "opacity": 1,
            "width": 60,
            "zIndex": 1,
          }
        }
      >
        <SvgXml
          style={
            Object {
              "color": "#FFF",
              "height": 18,
              "opacity": 0.4,
              "width": 18,
            }
          }
          xml="
  <svg width=\\"23\\" height=\\"24\\" viewBox=\\"0 0 23 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\">
    <path d=\\"M11.4854 3L2.22803 12.2573L11.4854 21.5146\\" stroke=\\"currentColor\\" stroke-opacity=\\"0.4\\" stroke-width=\\"3\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" />
  </svg>
"
        />
      </View>
      <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "#FFF",
            "fontFamily": "IBMPlexSans-Medium",
            "fontSize": 26,
            "letterSpacing": -0.3,
            "lineHeight": 34,
            "paddingHorizontal": 20,
            "position": "absolute",
            "textAlign": "center",
            "width": "100%",
          }
        }
      >
        Import Locations
      </Text>
    </View>
    <RCTScrollView
      style={
        Object {
          "alignSelf": "center",
          "flex": 1,
          "flexDirection": "column",
          "paddingLeft": 20,
          "paddingRight": 20,
          "textAlignVertical": "top",
          "width": "100%",
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "fontSize": 22,
              "fontWeight": "bold",
              "padding": 5,
              "paddingBottom": 20,
              "textAlign": "center",
            }
          }
        >
          <Text
            style={
              Object {
                "fontFamily": "IBMPlexSans",
                "fontSize": 16,
                "fontWeight": "normal",
                "lineHeight": 24,
                "marginTop": 12,
                "writingDirection": "ltr",
              }
            }
            use="body1"
          >
            Adding location data from Google will give you a head start on building your recent locations.
          </Text>
          <Text
            style={
              Object {
                "fontFamily": "IBMPlexSans",
                "fontSize": 16,
                "fontWeight": "normal",
                "lineHeight": 24,
                "marginTop": 12,
                "writingDirection": "ltr",
              }
            }
            use="body1"
          >
            Before you can import, you must first "Take out" your location data from Google.
          </Text>
          <Text
            style={
              Object {
                "fontFamily": "IBMPlexSans",
                "fontSize": 16,
                "fontWeight": "normal",
                "lineHeight": 24,
                "marginTop": 12,
                "writingDirection": "ltr",
              }
            }
            use="body1"
          >
            Visit Google Takeout and export your Location History using the following settings: 
1. Delivery method: "Add to Drive" 
2. Frequency: "Export once" 
3. File type & size: ".zip" and "1GB"
4. Google sends an email when the export is ready 
5. Return here to import locations from Google Drive
          </Text>
          <View
            accessibilityLabel="Visit Google Takeout"
            accessibilityRole="button"
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignContent": "center",
                "alignItems": "center",
                "alignSelf": "stretch",
                "borderBottomLeftRadius": 8,
                "borderBottomRightRadius": 8,
                "borderColor": "transparent",
                "borderStyle": "solid",
                "borderTopLeftRadius": 8,
                "borderTopRightRadius": 8,
                "borderWidth": 2,
                "flexDirection": "row",
                "height": 48,
                "justifyContent": "center",
                "marginTop": 24,
                "opacity": 1,
                "paddingHorizontal": 14,
              }
            }
            testID="google-takeout-link"
          >
            <Text
              small={true}
              style={
                Object {
                  "fontFamily": "IBMPlexSans-Medium",
                  "fontSize": 14,
                  "fontWeight": "normal",
                  "lineHeight": 21,
                  "writingDirection": "ltr",
                }
              }
              use="body1"
            >
              Visit Google Takeout
            </Text>
          </View>
          <View
            accessibilityLabel="Import Locations"
            accessibilityRole="button"
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignContent": "center",
                "alignItems": "center",
                "alignSelf": "stretch",
                "borderBottomLeftRadius": 8,
                "borderBottomRightRadius": 8,
                "borderColor": "transparent",
                "borderStyle": "solid",
                "borderTopLeftRadius": 8,
                "borderTopRightRadius": 8,
                "borderWidth": 2,
                "flexDirection": "row",
                "height": 48,
                "justifyContent": "center",
                "marginTop": 24,
                "opacity": 1,
                "paddingHorizontal": 14,
              }
            }
            testID="google-takeout-import-btn"
          >
            <Text
              small={true}
              style={
                Object {
                  "fontFamily": "IBMPlexSans-Medium",
                  "fontSize": 14,
                  "fontWeight": "normal",
                  "lineHeight": 21,
                  "writingDirection": "ltr",
                }
              }
              use="body1"
            >
              Import Locations
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </RCTSafeAreaView>,
]
`;
